Mathlib/RingTheory/Polynomial/Hermite/Basic.lean

```diff
@@ -0,0 +1,150 @@
+
+open Polynomial
+
+namespace Polynomial
+
+/-- the nth probabilist's Hermite polynomial -/
+noncomputable def hermite : ℕ → Polynomial ℤ
+  | 0 => 1
+  | n + 1 => X * hermite n - Polynomial.derivative (hermite n)
+#align polynomial.hermite Polynomial.hermite
+
+/-- The recursion `hermite (n+1) = (x - d/dx) (hermite n)` -/
+@[simp]
+theorem hermite_succ (n : ℕ) :
+    hermite (n + 1) = X * hermite n - Polynomial.derivative (hermite n) := by
+  rw [hermite]
+#align polynomial.hermite_succ Polynomial.hermite_succ
+
+theorem hermite_eq_iterate (n : ℕ) :
+    hermite n = ((fun p => X * p - Polynomial.derivative p)^[n]) 1 := by
+  induction' n with n ih
+  · rfl
+  · rw [Function.iterate_succ_apply', ← ih, hermite_succ]
+#align polynomial.hermite_eq_iterate Polynomial.hermite_eq_iterate
+
+@[simp]
+theorem hermite_zero : hermite 0 = C 1 :=
+  rfl
+#align polynomial.hermite_zero Polynomial.hermite_zero
+
+-- Porting note: There was initially @[simp] on this line but it was removed
+-- because simp can prove this theorem
+theorem hermite_one : hermite 1 = X := by
+  rw [hermite_succ, hermite_zero]
+  simp only [map_one, mul_one, derivative_one, sub_zero]
+#align polynomial.hermite_one Polynomial.hermite_one
+
+/-! ### Lemmas about `Polynomial.coeff` -/
+
+
+section Coeff
+
+theorem coeff_hermite_succ_zero (n : ℕ) : coeff (hermite (n + 1)) 0 = -coeff (hermite n) 1 := by
+  simp [coeff_derivative]
+#align polynomial.coeff_hermite_succ_zero Polynomial.coeff_hermite_succ_zero
+
+theorem coeff_hermite_succ_succ (n k : ℕ) :
+  coeff (hermite (n + 1)) (k + 1) = coeff (hermite n) k - (k + 2) * coeff (hermite n) (k + 2) := by
+  rw [hermite_succ, coeff_sub, coeff_X_mul, coeff_derivative, mul_comm]
+  norm_cast
+#align polynomial.coeff_hermite_succ_succ Polynomial.coeff_hermite_succ_succ
+
+theorem coeff_hermite_of_lt {n k : ℕ} (hnk : n < k) : coeff (hermite n) k = 0 := by
+  obtain ⟨k, rfl⟩ := Nat.exists_eq_add_of_lt hnk
+  clear hnk
+  induction' n with n ih generalizing k
+  · apply coeff_C
+  · have : n + k + 1 + 2 = n + (k + 2) + 1 := by ring
+    rw [Nat.succ_eq_add_one, coeff_hermite_succ_succ, add_right_comm, this, ih k, ih (k + 2),
+      MulZeroClass.mul_zero, sub_zero]
+#align polynomial.coeff_hermite_of_lt Polynomial.coeff_hermite_of_lt
+
+@[simp]
+theorem coeff_hermite_self (n : ℕ) : coeff (hermite n) n = 1 := by
+  induction' n with n ih
+  · apply coeff_C
+  · rw [coeff_hermite_succ_succ, ih, coeff_hermite_of_lt, MulZeroClass.mul_zero, sub_zero]
+    simp
+#align polynomial.coeff_hermite_self Polynomial.coeff_hermite_self
+
+-- Porting note: On line 116, the coercion from WithBot ℕ to ℕ was specified manually
+-- because it was not automatic
+@[simp]
+theorem degree_hermite (n : ℕ) : (hermite n).degree = n := by
+  rw [degree_eq_of_le_of_coeff_ne_zero]
+  simp_rw [degree_le_iff_coeff_zero, WithBot.coe_lt_coe]
+  · rintro m hnm
+    exact coeff_hermite_of_lt ((@WithBot.coe_lt_coe _ n m _).1 hnm)
+  · simp [coeff_hermite_self n]
```

Comment:
```suggestion
@[simp]
theorem degree_hermite (n : ℕ) : (hermite n).degree = n := by
  rw [degree_eq_of_le_of_coeff_ne_zero]
  simp_rw [degree_le_iff_coeff_zero]
   -- porting note: mathlib3 also had `simp_rw [WithBot.coe_lt_coe]` but it's not firing
   -- so we add it manually later
  · rintro m hnm
    exact coeff_hermite_of_lt (WithBot.coe_lt_coe.1 hnm)
  · simp [coeff_hermite_self n]
```
The real issue is that `simp_rw [WithBot.coe_lt_coe]` does nothing in mathlib4, for some reason; I've rewritten to give a more explicit explanation of the problem.